### PR TITLE
Context logging

### PR DESCRIPTION
**Pull request by jessepeterson**
Adds a simple `ctxlog` package which permits associating multiple `CtxKVFunc`s with a context. Then, when it's time to log, these are ran using the request context to pull out logging key-value pairs and create a new logger.

The end result is something like this for a typical NanoMDM push request.

Before:

```
2022/03/16 14:03:09 level=info handler=log addr=::1 method=GET path=/v1/push/D7083AF3-DCCB-5661-A678-BCE8F4D9A2C8 agent=curl/7.64.1
2022/03/16 14:03:09 level=info service=push msg=retrieved push cert topic=com.apple.mgmt.External.deadbeef-1f17-4c8e-8a63-dd17d3dd35d9
2022/03/16 14:03:09 level=debug handler=push msg=push count=1 errs=0
2022/03/16 14:03:09 level=info handler=log addr=192.168.0.2 method=PUT path=/mdm agent=MDM-OSX/1.0 mdmclient/1455 real_ip=2301:602:8a02:9b0f:422:25a8:481b:bd42
2022/03/16 14:03:09 level=info service=nanomdm status=Idle id=D7083AF3-DCCB-5661-A678-BCE8F4D9A2C8 type=Device
2022/03/16 14:03:09 level=debug service=nanomdm msg=no command retrieved id=D7083AF3-DCCB-5661-A678-BCE8F4D9A2C8
```

After:

```
2022/03/16 14:02:44 level=info handler=log trace_id=84172eff419691ca addr=::1 method=GET path=/v1/push/D7083AF3-DCCB-5661-A678-BCE8F4D9A2C8 agent=curl/7.64.1
2022/03/16 14:02:44 level=info service=push trace_id=84172eff419691ca msg=retrieved push cert topic=com.apple.mgmt.External.deadbeef-1f17-4c8e-8a63-dd17d3dd35d9
2022/03/16 14:02:44 level=debug handler=push trace_id=84172eff419691ca msg=push count=1 errs=0
2022/03/16 14:02:44 level=info handler=log trace_id=c35015253617eb34 addr=192.168.0.2 method=PUT path=/mdm agent=MDM-OSX/1.0 mdmclient/1455 real_ip=2301:602:8a02:9b0f:422:25a8:481b:bd42
2022/03/16 14:02:45 level=info service=nanomdm trace_id=c35015253617eb34 id=D7083AF3-DCCB-5661-A678-BCE8F4D9A2C8 type=Device status=Idle
2022/03/16 14:02:45 level=debug service=nanomdm trace_id=c35015253617eb34 id=D7083AF3-DCCB-5661-A678-BCE8F4D9A2C8 type=Device msg=no command retrieved
```

Note the `trace_id` value which tracks each log's request from its origin. Note also that `id` and `type` in the `service=nanomdm` layer is also a context value rather than explicitly logged now.

### CODE REVIEW

File: cmd/nanomdm/main.go [hunk 1]
```diff
@@ -196,6 +200,8 @@ func main() {
 		w.Write([]byte(`{"version":"` + version + `"}`))
 	})
 
+	rand.Seed(time.Now().UnixNano())
```
Comment (kilpatds) [inline cmd/nanomdm/main.go:203]: Is it possible to make this configurable for testing purposes?

**Review by kilpatds — APPROVED** [review]
Is there a way to up the log level for a specific request/other selector?

**Comment by jessepeterson** [comment]
@kilpatds re: log levels: I'm not sure if I'm reading you right, but there are only two log levels: Info and Debug. Wether those are logged or how is a detail left to the logger implementation and this `ctxlog` package just adapts a logger to add context (or not). The actual log line does happen until `.Info()` or `.Debug()` is called (where an implementation might decide to log or not). In NanoMDM's default case is handled by the `stdlogfmt` package whenever a Debug log is formatting:

https://github.com/micromdm/nanomdm/blob/6eb91ef1a894344461a1a048c3183e2794a727f8/log/stdlogfmt/stdlog.go#L43

For more details/philosophy on the simplistic logging "levels" see https://pkg.go.dev/github.com/jessepeterson/go-log#section-documentation and https://dave.cheney.net/2015/11/05/lets-talk-about-logging.

Hope that helps. :)

File: cmd/nanomdm/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (4e554c4c) [reply cmd/nanomdm/main.go:203]: main.go is the "reference" nanoMDM implementation afaik. Any other implementer who uses the NanoMDM library could do this in a separate way, so I don't think this is an issue.

[the remark this answers is not in the document]
Comment (kilpatds) [reply cmd/nanomdm/main.go:203]: It's a testability consideration, not anything else.

**Comment by kilpatds** [comment]
'''The log package should support fine grained control to enable or disable debug, and only debug, statements at the package or possibly even finer scope.'''

A pattern I've seen far too often around logging is something like
```
log.debug(context, <function that outputs 2k of text>())    
```
where it's a shame to generate 2k of text output just to throw it away because you don't actually log because debug logs aren't on. So ... that loop that provides context: if those functions are expensive, it's a shame to waste them on debug log lines that aren't actually enabled?  Or again, am I totally missing the boat?

File: cmd/nanomdm/main.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (kilpatds) [reply cmd/nanomdm/main.go:203]: (As far as I can tell, the only security implication of UUIDs is preventing collisions. So ... any seeding is "good enough", and use of a cryptographic PRNG is unnecessary. There's no harm if someone can guess a UUID.

So I'm just asking for "creation of unit test" purposes: if you can control the PRNG seed, you can get repeatable sequences out which makes testing easier.

If this isn't where anyone would actually test this, then ... yeah. Nevermind/drop)

**Comment by jessepeterson** [comment]
@kilpatds I think I see what you're saying now.

Yeah so the general pattern we follow is that we unconditionally call `logger.Debug()` and then the implementation decides whether to _actually_ log/output or whatever. I'm not aware of us generating huge amounts of debug code so we're just passing references to existing variables. That said: I do see the potential for a problem there. Perhaps the https://github.com/jessepeterson/go-log should add a simple `func IsDebug(): bool` interface so that you can conditionally do _debug log stuff_ only when you'll actually log it. Food for thought. However: I'd say that doing expensive debug stuff is different from _logging_ debug stuff: i.e. the module/package doing an expensive debug call that runs 2k of stuff ought to have its own debug mode, really. We don't really have anything doing expensive debug stuff (yet) though so none of our packages have separate debug modes — we just rely on the logger to do that for us (for now, at least).

Regarding computing the logger context stuff each invocation/being expensive: yes, I see that concern too and it's semi-addressed in two ways: one is that for some cases (Info logging) we only generated when we're going to log. The other cases, I'm just not really concerned about now.

In summary: yes, good points, I hear you, but not a concern and/or pre-mature optimization at this point. IMO, at least. :)
